File: 卡码/代码随想录/1/11-图论/Go/bfs/LeetCode_0827.go
// 827. 最大人工岛
package main

func max(a, b int) int {
	if a > b {
		return a
	} else {
		return b
	}
}

func largestIsland(grid [][]int) int {
	n, m := len(grid), len(grid[0])
	dir := [2][4]int{[4]int{0, 1, 0, -1}, [4]int{1, 0, -1, 0}}
	q := make([][2]int, 0)
	var bfs func(mark int) int
	bfs = func(mark int) int {
		cnt := 1
		for len(q) > 0 {
			v := q[0]
			q = q[1:]
			for i := 0; i < 4; i++ {
				nextX, nextY := v[0]+dir[0][i], v[1]+dir[1][i]
				if nextX < 0 || nextX > n-1 || nextY < 0 || nextY > m-1 {
					continue
				}
				if grid[nextX][nextY] == 0 || grid[nextX][nextY] > 1 {
					continue
				}
				q = append(q, [2]int{nextX, nextY})
				grid[nextX][nextY] = mark
				cnt++
			}
		}
		return cnt
	}
	mark := 2
	isAllLand := true
	areas := make(map[int]int)
	for i := 0; i < n; i++ {
		for j := 0; j < m; j++ {
			if grid[i][j] == 0 {
				isAllLand = false
			}
			if grid[i][j] == 1 {
				q = append(q, [2]int{i, j})
				grid[i][j] = mark
				cnt := bfs(mark)
				areas[mark] = cnt
				mark++
			}
		}
	}
	if isAllLand {
		return n * m
	}
	res := 0
	visited := make(map[int]bool)
	for i := 0; i < n; i++ {
		for j := 0; j < m; j++ {
			if grid[i][j] == 0 {
				cnt := 1
				visited = make(map[int]bool)
				for k := 0; k < 4; k++ {
					nextX, nextY := i+dir[0][k], j+dir[1][k]
					if nextX < 0 || nextX > n-1 || nextY < 0 || nextY > m-1 {
						continue
					}
					if _, ok := visited[grid[nextX][nextY]]; ok {
						continue
					}
					visited[grid[nextX][nextY]] = true
					cnt += areas[grid[nextX][nextY]]
				}
				res = max(res, cnt)
			}
		}
	}
	return res
}

File: 卡码/代码随想录/1/11-图论/Java/src/com/kurumi/bfs/LeetCode_0827.java
//827. 最大人工岛
package com.kurumi.bfs;

import java.util.ArrayDeque;
import java.util.Deque;
import java.util.HashMap;
import java.util.HashSet;
import java.util.Map.Entry;

public class LeetCode_0827 {

}
class Solution {
	int[][] dir=new int[][] {{0,1},{1,0},{0,-1},{-1,0}};
	Deque<Integer> q=new ArrayDeque<>();
	public int bfs(int[][] grid,int mark) {
		int cnt=1;
		while(!q.isEmpty()) {
			int x=q.poll();
			int y=q.poll();
			for(int i=0;i<4;i++) {
				int nextX=x+dir[i][0];
				int nextY=y+dir[i][1];
				if(nextX<0||nextX>grid.length-1||nextY<0||nextY>grid[0].length-1) continue;
				if(grid[nextX][nextY]==0||grid[nextX][nextY]>1) continue;
				q.offer(nextX);
				q.offer(nextY);
				cnt++;
				grid[nextX][nextY]=mark;
			}
		}
		return cnt;
	}
  public int largestIsland(int[][] grid) {
  	int n=grid.length;
  	int m=grid[0].length;
  	HashMap<Integer,Integer> areas = new HashMap<>();
  	int mark=2;
  	boolean isAllLand=true;
  	for(int i=0;i<n;i++) {
  		for(int j=0;j<m;j++) {
  			if(grid[i][j]==0) isAllLand=false;
  			if(grid[i][j]==1) {
  				q.offer(i);
  				q.offer(j);
  				grid[i][j]=mark;
  				int cnt=bfs(grid,mark);
  				areas.put(mark,cnt);
  				mark++;
  			}
  		}
  	}
  	if(isAllLand) return n*m;
  	int res=0;
  	HashSet<Integer> visited=new HashSet<>();
  	for(int i=0;i<n;i++) {
  		for(int j=0;j<m;j++) {
  			int count=1;
  			visited.clear();
  			if(grid[i][j]==0) {
  				for(int k=0;k<4;k++) {
  					int x=i+dir[k][0];
  					int y=j+dir[k][1];
  					if(x<0||x>n-1||y<0||y>m-1) continue;
  					if(grid[x][y]>1) {
  						if(visited.contains(grid[x][y])) continue;
  						count+=areas.get(grid[x][y]);
  						visited.add(grid[x][y]);
  					}
  				}
  			}
  			res=Math.max(res,count);
  		}
  	}
  	return res;
  }
}